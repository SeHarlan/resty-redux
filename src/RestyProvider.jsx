import React, { createContext, useContext, useEffect, useReducer } from 'react';
import PropTypes from 'prop-types';

import { customFetch } from './services/custom';

const RestyContext = createContext();

const initialState = {
  selectedOption: 'GET',
  json: '',
  url: '',
  name: '',
  password: '',
  bearerToken: '',
  history: [],
  fetchData: null,
  result: []
};
function reducer(state, action) {
  switch(action.type) {
    case 'setSelectedOption':
      return { ...state, selectedOption: action.payload };
    case 'setJson':
      return { ...state, json: action.payload };
    case 'setUrl': 
      return { ...state, url: action.payload };
    case 'setName':
      return { ...state, name: action.payload };
    case 'setPassword':
      return { ...state, password: action.payload };
    case 'setBearerToken':
      return { ...state, bearerToken: action.payload };
    case 'setHistory':
      return { ...state, history: [...state.history, action.payload] };
    case 'setFetchData': 
      return { ...state, fetchData: action.payload };
    case 'setResult':
      return { ...state, result: action.payload };
    default:
      return state;
  }
}

const RestyProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const { fetchData } = state;

  useEffect(() => {
    if(!fetchData) return; 
    dispatch({ type: 'setResult', payload: ['...Loading (or Nothing Found)'] });
    customFetch(fetchData).then(res => {
      if(res === []) {
        dispatch({ type: 'setResult', payload: ['No Results Found'] });
      } else {
        dispatch({ type: 'setResult', payload: res });
        dispatch({ type: 'setHistory', payload: fetchData });
      }
    });
  }, [fetchData]);

  return (
    <RestyContext.Provider value={{ state, dispatch }}>
      {children}
    </RestyContext.Provider>
  );
};

RestyProvider.propTypes = {
  children: PropTypes.node
};

export default RestyProvider;

export const useGlobalState = () => {
  const { state } = useContext(RestyContext);
  return state;
};
export const useDispatch = () => {
  const { dispatch } = useContext(RestyContext);
  return dispatch;
};

export const useHandleSubmit = (e) => {
  e.preventDefault();
  const { url, selectedOption, json, name, password, bearerToken } = useGlobalState();
  const dispatch = useDispatch();
  dispatch({ type: 'setFetchData', payload: ({
    url: url,
    method: selectedOption,
    json: json,
    username: name,
    password: password,
    bearerToken: bearerToken
  })
  });
};
export const useHandleChange = ({ target }, toChange) => {
  const dispatch = useDispatch();
  dispatch({ type: toChange, payload: target.value });
};

