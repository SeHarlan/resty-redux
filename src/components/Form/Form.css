.Form {
  background-color: seashell;
  padding: 1rem;
  {}
  input[type=text] {
    width: 100%;
    padding: 5px;
  }
  section {
    margin: 1rem;
  }
  label {
    margin: 1rem ;
    padding: 0.5rem;
    border-radius: 0.2rem;
    box-shadow: inset 0 0 1rem white;
  }
  textarea {
    width: 100%;
  }
  input[type=radio] {
    display: none;
  }
  input[type=radio]:checked + label {
    background-color: sandybrown;
    box-shadow: inset 0 0 0.4rem rgba(0, 0, 0, 0.3);

  }
  button {
    width: 60px;
    height: 30px;
    border-radius: 0.5rem;
    background-color: burlywood;
    outline: none;
    margin: 0.5rem;
  }
}

.Headers {
  display: grid;
  grid-template-columns: 1fr 1fr;
  align-items: center;
  text-align: right;
  {}
  h3 {
    grid-column: 1 / span 2;
    margin: 0;
    text-align: center;
    text-decoration: underline;
  }
  span {
    margin: 0.5rem;
  }
  input {
    margin: 0.5rem;
  }
}
.Json {
  display: grid;
  grid-template-columns: 1fr 1fr;
}