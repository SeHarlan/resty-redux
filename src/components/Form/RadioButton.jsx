import React from 'react';
import PropTypes from 'prop-types';

const RadioButton = ({ value, selectedOption, onOptionChange }) => (
  <>
    <input checked={selectedOption === value} id={value} type="radio" name="method" value={value} onChange={onOptionChange} />
    <label htmlFor={value}>{value}</label>
  </>
);
RadioButton.propTypes = {
  selectedOption: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  onOptionChange: PropTypes.func.isRequired
};
export default RadioButton;
