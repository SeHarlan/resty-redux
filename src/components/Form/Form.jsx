import React from 'react'; 
import RadioButton from './RadioButton';
import style from './Form.css';
import { useGlobalState, useHandleSubmit, useDispatch } from '../../RestyProvider';


const Form = () => {
  const { 
    url: urlInput, 
    json: jsonInput, 
    name: usernameInput, 
    password: passwordInput, 
    bearerToken: bearerTokenInput, 
    selectedOption = 'GET', 
  } = useGlobalState();

  const dispatch = useDispatch();
  const HandleChange = ({ target }, toChange) => {
    dispatch({ type: toChange, payload: target.value });
  };
  const HandleSubmit = (e) => {
    e.preventDefault();

    dispatch({ type: 'setFetchData', payload: ({
      url: urlInput,
      method: selectedOption,
      json: jsonInput,
      username: usernameInput,
      password: passwordInput,
      bearerToken: bearerTokenInput
    })
    });
  };
  

  const radios = ['GET', 'POST', 'PUT', 'PATCH', 'DELETE'].map(method => (
    <RadioButton key={method} value={method} selectedOption={selectedOption} onOptionChange={(e) => HandleChange(e, 'setSelectedOption')} />));

  return (
    <form className={style.Form} onSubmit={HandleSubmit}>

      <input type="text" name="url" value={urlInput} onChange={(e) => HandleChange(e, 'setUrl')} placeholder="URL"/>
      <section>

        {radios}
        <button type="submit">Fetch</button>
      </section>
      <section className={style.Json}>
        <textarea name="json" value={jsonInput} cols="30" rows="10" onChange={(e) => HandleChange(e, 'setJson')} placeholder="Raw JSON body"></textarea>
        <div className={style.Headers}>
          <h3>Basic Authorization</h3>
          <span>Username: </span><input type="text" name="username" value={usernameInput} onChange={(e) => HandleChange(e, 'setName')} placeholder="Username"/>
      
          <span>Password: </span><input type="text" name="password" value={passwordInput} onChange={(e) => HandleChange(e, 'setPassword')} placeholder="Password"/>
        
          <span>Bearer Token: </span><input type="text" name="bearerToken" value={bearerTokenInput} onChange={(e) => HandleChange(e, 'setBearerToken')} placeholder="Bearer Token" />
        </div>
      </section>
    </form>);
};

export default Form;
