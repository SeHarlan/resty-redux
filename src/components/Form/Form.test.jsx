import React from 'react';
import { shallow } from 'enzyme';
import Form from './Form';
import RestyProvider from '../../RestyProvider';

describe('Form component', () => {
  it('renders Form', () => {
    const wrapper = shallow(
      <RestyProvider>
        <Form />
      </RestyProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
