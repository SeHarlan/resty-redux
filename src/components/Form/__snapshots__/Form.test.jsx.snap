// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component renders Form 1`] = `
<ContextProvider
  value={
    Object {
      "dispatch": [Function],
      "state": Object {
        "bearerToken": "",
        "fetchData": null,
        "history": Array [],
        "json": "",
        "name": "",
        "password": "",
        "result": Array [],
        "selectedOption": "GET",
        "url": "",
      },
    }
  }
>
  <Form />
</ContextProvider>
`;
